File: tests/integration/scripts/combined_test.py
#!/usr/bin/env python3
"""
Combined test script that exercises filesystem, network, and process operations.
This simulates a realistic agent workload.
"""
import os
import sys
import json
import subprocess
import tempfile
import time

# Try to use requests
try:
    import requests
    HAS_REQUESTS = True
except ImportError:
    HAS_REQUESTS = False
    import urllib.request

def main():
    print("=== Combined Agent Simulation ===")
    print(f"PID: {os.getpid()}")
    start_time = time.time()

    # Phase 1: Set up working directory
    print("\n[Phase 1] Setting up workspace...")
    workspace = tempfile.mkdtemp(prefix="agent_workspace_")
    print(f"  Workspace: {workspace}")

    # Phase 2: Download some data (network)
    print("\n[Phase 2] Fetching external data...")
    try:
        if HAS_REQUESTS:
            resp = requests.get("https://httpbin.org/json", timeout=10)
            data = resp.json()
        else:
            with urllib.request.urlopen("https://httpbin.org/json", timeout=10) as resp:
                data = json.loads(resp.read().decode())
        print(f"  Fetched JSON with {len(data)} keys")
    except Exception as e:
        print(f"  Network fetch failed: {e}")
        data = {"fallback": True, "reason": str(e)}

    # Phase 3: Write data to file (filesystem)
    print("\n[Phase 3] Saving data to disk...")
    data_file = os.path.join(workspace, "fetched_data.json")
    with open(data_file, 'w') as f:
        json.dump(data, f, indent=2)
    print(f"  Saved to: {data_file}")

    # Phase 4: Process data with subprocess (process)
    print("\n[Phase 4] Processing data...")
    result = subprocess.run(
        ["wc", "-c", data_file],
        capture_output=True,
        text=True
    )
    byte_count = result.stdout.split()[0]
    print(f"  File size: {byte_count} bytes")

    # Phase 5: Create analysis output
    print("\n[Phase 5] Generating analysis...")
    analysis = {
        "timestamp": time.time(),
        "workspace": workspace,
        "data_file": data_file,
        "file_size_bytes": int(byte_count),
        "network_success": "fallback" not in data,
        "duration_ms": (time.time() - start_time) * 1000
    }
    analysis_file = os.path.join(workspace, "analysis.json")
    with open(analysis_file, 'w') as f:
        json.dump(analysis, f, indent=2)
    print(f"  Analysis saved to: {analysis_file}")

    # Phase 6: Run a subprocess that does its own file I/O
    print("\n[Phase 6] Running nested subprocess...")
    script = f'''
import json
with open("{analysis_file}") as f:
    data = json.load(f)
print(f"Nested process read {{len(data)}} analysis fields")
with open("{os.path.join(workspace, 'nested_output.txt')}", 'w') as f:
    f.write("Output from nested process\\n")
'''
    result = subprocess.run(
        ["python3", "-c", script],
        capture_output=True,
        text=True
    )
    print(f"  {result.stdout.strip()}")

    # Verify nested output exists
    nested_output = os.path.join(workspace, "nested_output.txt")
    if os.path.exists(nested_output):
        print(f"  Nested output file created: {nested_output}")

    # Phase 7: Cleanup
    print("\n[Phase 7] Cleanup...")
    for f in os.listdir(workspace):
        os.remove(os.path.join(workspace, f))
    os.rmdir(workspace)
    print(f"  Removed workspace: {workspace}")

    total_time = (time.time() - start_time) * 1000
    print(f"\n=== Combined Test Complete ({total_time:.0f}ms) ===")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: tests/integration/scripts/combined_test.js
#!/usr/bin/env node
/**
 * Combined test script that exercises filesystem, network, and process operations.
 * This simulates a realistic agent workload.
 */
const fs = require('fs');
const path = require('path');
const os = require('os');
const https = require('https');
const { execSync, spawn } = require('child_process');

function fetch(url) {
    return new Promise((resolve, reject) => {
        const req = https.get(url, { timeout: 10000 }, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => resolve({ status: res.statusCode, body: data }));
        });
        req.on('error', reject);
        req.on('timeout', () => {
            req.destroy();
            reject(new Error('Request timeout'));
        });
    });
}

async function main() {
    console.log("=== Combined Agent Simulation (Node.js) ===");
    console.log(`PID: ${process.pid}`);
    const startTime = Date.now();

    // Phase 1: Set up working directory
    console.log("\n[Phase 1] Setting up workspace...");
    const workspace = fs.mkdtempSync(path.join(os.tmpdir(), 'agent_workspace_node_'));
    console.log(`  Workspace: ${workspace}`);

    // Phase 2: Download some data (network)
    console.log("\n[Phase 2] Fetching external data...");
    let data;
    try {
        const { body } = await fetch("https://httpbin.org/json");
        data = JSON.parse(body);
        console.log(`  Fetched JSON with ${Object.keys(data).length} keys`);
    } catch (err) {
        console.log(`  Network fetch failed: ${err.message}`);
        data = { fallback: true, reason: err.message };
    }

    // Phase 3: Write data to file (filesystem)
    console.log("\n[Phase 3] Saving data to disk...");
    const dataFile = path.join(workspace, 'fetched_data.json');
    fs.writeFileSync(dataFile, JSON.stringify(data, null, 2));
    console.log(`  Saved to: ${dataFile}`);

    // Phase 4: Process data with subprocess (process)
    console.log("\n[Phase 4] Processing data...");
    const wcOutput = execSync(`wc -c "${dataFile}"`, { encoding: 'utf-8' });
    const byteCount = wcOutput.trim().split(/\s+/)[0];
    console.log(`  File size: ${byteCount} bytes`);

    // Phase 5: Create analysis output
    console.log("\n[Phase 5] Generating analysis...");
    const analysis = {
        timestamp: Date.now(),
        workspace: workspace,
        data_file: dataFile,
        file_size_bytes: parseInt(byteCount),
        network_success: !data.fallback,
        duration_ms: Date.now() - startTime
    };
    const analysisFile = path.join(workspace, 'analysis.json');
    fs.writeFileSync(analysisFile, JSON.stringify(analysis, null, 2));
    console.log(`  Analysis saved to: ${analysisFile}`);

    // Phase 6: Run a subprocess that does its own file I/O
    console.log("\n[Phase 6] Running nested subprocess...");
    const nestedOutput = path.join(workspace, 'nested_output.txt');
    const script = `
        const fs = require('fs');
        const data = JSON.parse(fs.readFileSync('${analysisFile}'));
        console.log('Nested process read ' + Object.keys(data).length + ' analysis fields');
        fs.writeFileSync('${nestedOutput}', 'Output from nested process\\n');
    `;
    const result = execSync(`node -e "${script.replace(/"/g, '\\"')}"`, { encoding: 'utf-8' });
    console.log(`  ${result.trim()}`);

    if (fs.existsSync(nestedOutput)) {
        console.log(`  Nested output file created: ${nestedOutput}`);
    }

    // Phase 7: Cleanup
    console.log("\n[Phase 7] Cleanup...");
    for (const f of fs.readdirSync(workspace)) {
        fs.unlinkSync(path.join(workspace, f));
    }
    fs.rmdirSync(workspace);
    console.log(`  Removed workspace: ${workspace}`);

    const totalTime = Date.now() - startTime;
    console.log(`\n=== Combined Test Complete (${totalTime}ms) ===`);
    return 0;
}

main().then(process.exit).catch(err => {
    console.error(err);
    process.exit(1);
});
